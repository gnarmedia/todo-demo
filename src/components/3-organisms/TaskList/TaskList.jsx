// import dependencies
import React, { Fragment, Component } from "react";
import PropTypes from "prop-types";

// import components
import Task from "../Task/Task.jsx";
import AddTask from "../AddTask/AddTask.jsx";
import FilterTask from "../FilterTask/FilterTask.jsx";

/**
 * @name TaskList
 * @desc A TaskList class with a list of tasks with statuses
 */
class TaskList extends Component {
    /**
     * @name constructor
     * @param {object} props (tasks and filter)
     */
    constructor(props) {
        super(props);
        this.state = {
            filter: props.filter,
            tasks: props.tasks
        };

        this._renderTask = this._renderTask.bind(this);
        this.handleToggleStatusClick = this.handleToggleStatusClick.bind(this);
        this.handleAddSubmit = this.handleAddSubmit.bind(this);
        this.handleRemoveClick = this.handleRemoveClick.bind(this);
        this.handleFilterClick = this.handleFilterClick.bind(this);
    }

    /**
     * @name _renderTask
     * @desc task render helper
     * @param {object} task
     * @returns {function} Task component with li
     */
    _renderTask({ id, title, isComplete }) {
        return (
            <li key={id.toString()}>
                <Task
                    title={title}
                    isComplete={isComplete}
                    onToggleStatusClick={() => this.handleToggleStatusClick(id)}
                    onRemoveClick={() => this.handleRemoveClick(id)}
                />
            </li>
        );
    }

    /**
     * @name _filterTask
     * @desc Determines if task should be filtered or not
     * @param {string} filter
     * @param {object} task
     * @returns {boolean} Flag for whether or not task should be filtered
     */
    _shouldFilterTask(filter, task) {
        if (
            (filter === "COMPLETE" && !task.isComplete) ||
            (filter === "INCOMPLETE" && task.isComplete)
        ) {
            return false;
        }

        return true;
    }

    /**
     * @name handleToggleStatusClick
     * @desc Handles the click event for the task status toggle
     * @param {striing} id
     */
    handleToggleStatusClick(id) {
        this.setState((prevState, props) => {
            const tasksMapped = prevState.tasks.map(task => {
                if (task.id === id) {
                    task.isComplete = !task.isComplete;
                }

                return task;
            });

            localStorage.setItem("tasks", JSON.stringify(tasksMapped));
            return tasksMapped;
        });
    }

    /**
     * @name handleAddSubmit
     * @desc Handles submission of task to add
     * @param {object} task
     */
    handleAddSubmit(task) {
        this.setState((prevState, props) => {
            const tasks = Object.assign({}, prevState).tasks;

            tasks.splice(0, 0, task);

            localStorage.setItem("tasks", JSON.stringify(tasks));
            return { tasks };
        });
    }

    /**
     * @name handleRemoveClick
     * @desc Handles the click event for the remove task button
     * @param {*} id 
     */
    handleRemoveClick(id) {
        this.setState((prevState, props) => {
            const tasks = Object.assign({}, prevState).tasks.filter(
                x => x.id !== id
            );

            localStorage.setItem("tasks", JSON.stringify(tasks));
            return { tasks };
        });
    }

    /**
     * @name handleFilterClick
     * @desc Handles the click event for the task filter buttons
     * @param {string} filter 
     */
    handleFilterClick(filter) {
        localStorage.setItem("filter", JSON.stringify(filter));
        this.setState((prevState, props) => ({ filter }));
    }

    /**
     * @name render
     * @desc Renders TaskList component
     * @returns {function} TaskList component
     */
    render() {
        const filter = this.state.filter;

        return (
            <Fragment>
                <AddTask onSubmit={this.handleAddSubmit} />
                <ul>
                    {this.state.tasks
                        .filter(task => this._shouldFilterTask(filter, task))
                        .map(this._renderTask)}
                </ul>
                <FilterTask onFilterClick={this.handleFilterClick} />
            </Fragment>
        );
    }
}

// These are proptypes, they are used to define the expected minimum prop values
// for a component
TaskList.propTypes = {
    filter: PropTypes.string.isRequired,
    tasks: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            isComplete: PropTypes.bool.isRequired,
            title: PropTypes.string.isRequired
        }).isRequired
    ).isRequired
};

// export our component for importing into other modules
export default TaskList;
