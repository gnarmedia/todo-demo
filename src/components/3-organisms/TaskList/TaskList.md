```jsx
const tasks = [
    { id: "ryBRDzBdM", title: "Task 1", isComplete: false },
    { id: "HygBRPGBuG", title: "Task 2", isComplete: true },
    { id: "r1ZSCDMruM", title: "Task 3", isComplete: false }
];
<TaskList tasks={tasks} filter="ALL" />
```