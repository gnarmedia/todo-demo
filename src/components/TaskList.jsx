// import dependencies
import React from "react";
import PropTypes from "prop-types";

// import components
import Task from "./Task.jsx";

/**
 * @name: _renderTask
 * @desc: This function renders the task component
 * @param {object} task
 */
const _renderTask = ({ id, title, isComplete }) => (
    <li key={id.toString()}>
        <Task title={title} isComplete={isComplete} />
    </li>
);

/**
 * @name: TaskList
 * @desc: A list of tasks with statuses
 * @param: {array} tasks
 * @returns: {function}
 */
const TaskList = ({ tasks }) => {
    return <ul>{tasks.map(_renderTask)}</ul>;
};

// These are proptypes, they are used to define the expected minimum prop values
// for a component
TaskList.propTypes = {
    tasks: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            isComplete: PropTypes.bool.isRequired,
            title: PropTypes.string.isRequired
        }).isRequired
    ).isRequired
};

// export our component for importing into other modules
export default TaskList;
