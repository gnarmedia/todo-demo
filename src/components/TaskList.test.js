import React from "react";
import renderer from "react-test-renderer";

// import components
import TaskList from "./TaskList.jsx";

describe("TaskList", () => {
    it("should render the TaskList", () => {
        const tasks = [
                { id: "1", title: "Task 1", isComplete: false },
                { id: "2", title: "Task 2", isComplete: true },
                { id: "3", title: "Task 3", isComplete: false }
            ],
            tree = renderer.create(<TaskList tasks={tasks} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
