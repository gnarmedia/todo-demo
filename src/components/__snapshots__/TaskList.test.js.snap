// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList should render the TaskList 1`] = `
<ul>
  <li>
    <input
      defaultChecked={false}
      type="checkbox"
    />
    Task 1
     [
    incomplete
    ]
  </li>
  <li>
    <input
      defaultChecked={true}
      type="checkbox"
    />
    Task 2
     [
    complete
    ]
  </li>
  <li>
    <input
      defaultChecked={false}
      type="checkbox"
    />
    Task 3
     [
    incomplete
    ]
  </li>
</ul>
`;
